types/signatures.go

```diff
@@ -407,17 +417,72 @@ func (uh UnlockHash) MarshalJSON() ([]byte, error) {
 // that has been encoded to a hex string.
 func (uh *UnlockHash) UnmarshalJSON(b []byte) error {
 	// Check the length of b.
-	if len(b) != crypto.HashSize*2+2 {
+	if len(b) != crypto.HashSize*2+UnlockHashChecksumSize*2+2 {
+		return ErrUnlockHashWrongLen
+	}
+
+	// Decode the unlock hash.
+	var byteUnlockHash []byte
+	var checksum []byte
+	_, err := fmt.Sscanf(string(b[1:1+crypto.HashSize*2]), "%x", &byteUnlockHash)
+	if err != nil {
+		return err
+	}
+	// Decode the checksum.
+	_, err = fmt.Sscanf(string(b[1+crypto.HashSize*2:1+crypto.HashSize*2+UnlockHashChecksumSize*2]), "%x", &checksum)
+	if err != nil {
+		return err
+	}
+
+	// Verify the checksum - leave uh as-is unless the checksum is valid.
+	var unlockHash UnlockHash
+	copy(unlockHash[:], byteUnlockHash)
+	expectedChecksum := crypto.HashObject(unlockHash)
+	if bytes.Compare(expectedChecksum[:UnlockHashChecksumSize], checksum) != 0 {
+		return ErrInvalidUnlockHashChecksum
+	}
+	copy(uh[:], unlockHash[:])
+
+	return nil
+}
+
+// String returns the hex representation of the unlock hash as a string - this
+// includes a checksum.
+func (uh UnlockHash) String() string {
+	uhChecksum := crypto.HashObject(uh)
+	return fmt.Sprintf("%x%x", uh[:], uhChecksum[:UnlockHashChecksumSize])
+}
+
+// LoadString loads a hex representation (including checksum) of an unlock hash
+// into an unlock hash object. An error is returned if the string is invalid or
+// fails the checksum.
+func (uh *UnlockHash) LoadString(strUH string) error {
+	// Check the length of strUH.
+	if len(strUH) != crypto.HashSize*2+UnlockHashChecksumSize*2 {
 		return ErrUnlockHashWrongLen
 	}
 
-	// Decode the input .
+	// Decode the unlock hash.
 	var byteUnlockHash []byte
-	_, err := fmt.Sscanf(string(b[1:crypto.HashSize*2+1]), "%x", &byteUnlockHash)
+	var checksum []byte
+	_, err := fmt.Sscanf(strUH[:crypto.HashSize*2], "%x", &byteUnlockHash)
+	if err != nil {
+		return err
+	}
+	// Decode the checksum.
+	_, err = fmt.Sscanf(strUH[crypto.HashSize*2:crypto.HashSize*2+UnlockHashChecksumSize*2], "%x", &checksum)
```

Comment:
lots of duplicated code here. Either `UnmarshalJSON` should call `LoadString`, or the other way around.

Really we should implement a `Scan` method so that we can use `fmt.Scan` directly, but it's more trouble than it's worth right now.